### PR TITLE
Add web-vitals integration

### PR DESCRIPTION
**Pull request by delucis**
## Changes

- Adds a new `@astrojs/web-vitals` integration
- ~~This branch currently includes some changes to `@astrojs/db` from #10882 that will go away once that is merged~~

## Testing

Added some basic tests for dev mode. Not quite sure what a build test would look like or if we should have end-to-end tests for the client-side script?

## Docs

Minimal docs in the README for now cc @withastro/maintainers-docs

### CODE REVIEW

File: packages/integrations/web-vitals/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+							'Please add an SSR adapter: https://docs.astro.build/en/guides/server-side-rendering/'
+					);
+				}
+
+				// Middleware that adds a `<meta>` tag to each page.
+				addMiddleware({ entrypoint: '@astrojs/web-vitals/middleware', order: 'post' });
+				// Endpoint that collects metrics and inserts them in Astro DB.
+				injectRoute({
+					entrypoint: '@astrojs/web-vitals/endpoint',
+					pattern: '/_/astro-vitals',
```
Comment (delucis) [inline packages/integrations/web-vitals/src/index.ts:33]: Currently injecting the data collection endpoint to `/_/astro-vitals/`.

Should we discuss whether we want a default namespacing approach? For example, the image service uses `/_image`.

File: packages/integrations/web-vitals/src/middleware.ts
Comment (delucis) [inline packages/integrations/web-vitals/src/middleware.ts:1]: This middleware preserves streaming and injects a `<meta>` tag to each page’s `<head>`. Wouldn’t mind someone taking a quick look to make sure there are no footguns here.

**Review by delucis — COMMENTED** [review]
Ready for review! Testing is a little bit minimal. Not sure if we’d want an end-to-end test for the client script.

File: packages/integrations/web-vitals/src/env.d.ts
Comment (bluwy) [inline packages/integrations/web-vitals/src/env.d.ts:1]: Is this file intended to be committed?

**Review by bluwy — APPROVED** [review]
Looks great! The logic seems good to me.

File: packages/integrations/web-vitals/src/index.ts (hunk 1 above)
Comment (bluwy) [reply packages/integrations/web-vitals/src/index.ts:33]: `/_/` sounds nice, I suppose we can only change `/_image` to it in a major since it's somewhat exposed?

Also bikeshedding 😅 Should we drop the `astro` and do `/_/web-vitals` or `/_/vitals` instead?

File: packages/integrations/web-vitals/src/env.d.ts
Comment (delucis) [reply packages/integrations/web-vitals/src/env.d.ts:1]: Yes — currently Astro DB types aren’t working great for integrations that don’t have an Astro project environment to refer to, so we’re kind of working around that with this triple-slash reference.

Hopefully something @bholmesdev or a TS magician might be able to improve in the future?

File: packages/integrations/web-vitals/src/index.ts (hunk 1 above)
Comment (delucis) [reply packages/integrations/web-vitals/src/index.ts:33]: Yeah, there are a bunch of potential approaches. I guess most extreme would be vendor-style namespacing matching the package name:

```
/_/astrojs/web-vitals/
```

Or alternatively, we don’t worry too much, and eventually support users customising the endpoint location to handle any conflicts.

File: packages/integrations/web-vitals/src/client-script.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		pathname,
+		route,
+		name,
+		id,
+		value,
+		rating,
+	}));
+	const body = JSON.stringify(rawBody);
+	const endpoint = '/_/astro-vitals';
+	if (navigator.sendBeacon) navigator.sendBeacon(endpoint, body);
```
Comment (bholmesdev) [inline packages/integrations/web-vitals/src/client-script.ts:24]: TIL about `sendBeacon` 💡

File: packages/integrations/web-vitals/src/index.ts (hunk 1 above)
Comment (bholmesdev) [reply packages/integrations/web-vitals/src/index.ts:33]: Naming is always tough! Maybe we could match the namespacing expected by `astro add`:
- use vendor prefixing for third party packages
- drop the "astro" for first party packages

So use either `/_/web-vitals/` or `/_web-vitals/`. I bias towards the latter to match `_images` and avoid introducing a new competing standard.

File: packages/integrations/web-vitals/src/middleware.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	route = miniEncodeAttribute(stripTrailingSlash(route));
+	return `<meta name="x-astro-vitals-route" content="${route}" />`;
+}
+
+function stripTrailingSlash(str: string) {
+	return str.length > 1 && str.at(-1) === '/' ? str.slice(0, -1) : str;
+}
+
+function miniEncodeAttribute(str: string) {
```
Comment (bholmesdev) [inline packages/integrations/web-vitals/src/middleware.ts:54]: We sure this covers all necessary cases?

**Review by bholmesdev — APPROVED** [review]
Implementation looks robust and well-tested. A `vital` addition to Astro

File: packages/integrations/web-vitals/src/index.ts (hunk 1 above)
Comment (delucis) [reply packages/integrations/web-vitals/src/index.ts]: Thanks for the feedback! Renamed to `/_web-vitals` and moved to a shared constant, so we can easily rename in the future if we want.

File: packages/integrations/web-vitals/src/middleware.ts (hunk 3 above)
Comment (delucis) [reply packages/integrations/web-vitals/src/middleware.ts:54]: I’m pretty sure for attributes it does? I think in fact even `<` and `>` would be safe unencoded — just tried in an Astro component and this input:
```jsx
content={'/blog/&"<>\\\'/'}
```
output to
```html
content="/blog/&#38;&#34;<>\'/"
```
So only the `&` and the `"` were escaped.
